back/engines/commercial/report_builder/spec/services/report_builder/queries/visitors_spec.rb

```diff
@@ -0,0 +1,276 @@
+        {
+          visits: 2,
+          visitors: 1,
+          date_group: Date.new(2022, 10, 10)
+        }
+      ])
+    end
+
+    it 'filters by project' do
+      params = {
+        start_at: Date.new(2022, 8, 1),
+        end_at: Date.new(2022, 11, 1),
+        project_id: @project_id
+      }
+
+      expect(query.run_query(**params)).to eq({
+        visitors_timeseries: [
+          {
+            visits: 2,
+            visitors: 2,
+            date_group: Date.new(2022, 9, 1)
+          }
+        ],
+        visits_whole_period: 2,
+        visitors_whole_period: 2,
+        avg_seconds_per_session_whole_period: 240,
+        avg_pages_visited_whole_period: 2
+      })
+    end
+
+    it 'returns correct data with compared period' do
+      params = {
+        start_at: Date.new(2022, 10, 1),
+        end_at: Date.new(2022, 11, 1),
+        compare_start_at: Date.new(2022, 9, 1),
+        compare_end_at: Date.new(2022, 10, 1)
+      }
+
+      expect(query.run_query(**params)).to eq({
+        visitors_timeseries: [
+          {
+            visits: 5,
+            visitors: 2,
+            date_group: Date.new(2022, 10, 1)
+          }
+        ],
+        visits_whole_period: 5,
+        visitors_whole_period: 2,
+        avg_seconds_per_session_whole_period: 60 * 1.2,
+        avg_pages_visited_whole_period: 1.2,
+
+        visits_compared_period: 3,
+        visitors_compared_period: 2,
+        avg_seconds_per_session_compared_period: 200,
+        avg_pages_visited_compared_period: 2
+      })
+    end
+
+    it 'works if everything is nil' do
+      params = {
+        start_at: Date.new(2023, 10, 1),
+        end_at: Date.new(2023, 11, 1)
+      }
+
+      expect(query.run_query(**params)).to eq({
+        visitors_timeseries: [],
+        visits_whole_period: 0,
+        visitors_whole_period: 0,
+        avg_seconds_per_session_whole_period: 0,
+        avg_pages_visited_whole_period: 0
+      })
+    end
+
+    it 'applies exclude_roles filter' do
+      # Create session december (admin)
+      session = create(:session, created_at: Date.new(2022, 12, 2), highest_role: 'admin')
+      create(:pageview, session_id: session.id, path: '/en/', created_at: DateTime.new(2022, 12, 2, 10, 0, 0))
+
+      params = {
+        start_at: Date.new(2022, 8, 1),
+        end_at: Date.new(2023, 1, 1),
+        exclude_roles: ['admin']
+      }
+
+      expect(query.run_query(**params)).to eq({
+        visitors_timeseries: [
+          {
+            visits: 3,
+            visitors: 2,
+            date_group: Date.new(2022, 9, 1)
+          },
+          {
+            visits: 5,
+            visitors: 2,
+            date_group: Date.new(2022, 10, 1)
+          }
+        ],
+        visits_whole_period: 8,
+        visitors_whole_period: 4,
+        avg_seconds_per_session_whole_period: 135,
+        avg_pages_visited_whole_period: 1.5
+      })
+    end
+
+    it 'removes page views where time spent on page is over 2 hours' do
```

Comment:
I figured that this was desirable- imagine that someone leaves their computer open for a few days and then clicks on the next page. This could significantly change the average session time. So I thought, better to ignore it when people spend more than 2 hours on a page

Reply:
Makes sense